### PR TITLE
RFC: Principles for Generating Reusable Code 

### PR DESCRIPTION
**Pull request by hlbarber**
[Rendered](https://github.com/awslabs/smithy-rs/blob/810c8cdc1a6a90da3a07d2ffc5720b26dc717d9b/design/src/rfcs/rfc0034_generating_reusable_code.md)

### CODE REVIEW

File: design/src/rfcs/rfc0034_generating_reusable_code.md [hunk 1]
```diff
@@ -0,0 +1,232 @@
+# RFC: Principles for Generating Reusable Code
+
+> Status: Accepted
```
Comment (82marbag) [inline design/src/rfcs/rfc0034_generating_reusable_code.md:3]: On a previous comment: @LukeMathWalker I'd be happy to accept for merge interesting RFCs, regardless of whether they're accepted or not

Comment (LukeMathWalker) [reply design/src/rfcs/rfc0034_generating_reusable_code.md:3]: I disagree - we can merge an RFC either as `Accepted` or `Rejected`. Why would we merge an RFC without a clear outcome? 🤔

File: design/src/rfcs/rfc0034_generating_reusable_code.md [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+This RFC is does _not_ propose specific changes to the code generators API or the client or server Rust API. However the principles, when applied, are intended to enable desirable APIs for both the code generator and the client and server SDK.
+
+## Terminology
+
+- **Compatible types**: Two types are said to be "compatible", if they could in principle be collapsed into a single type. That is, if they have the identical type definition and non-conflicting methods/trait implementations.
+- **Abstract Syntax Tree (AST)**: A tree representing the _syntactic_ structure of the source code of a formal language.
+- **Abstract Semantic Graph (ASG)**: A graph representing the _semantic_ structure of the source code of a formal language. In our context, one can think of an ASG as an AST with additional edges introduced between nodes by the presence of identifiers. For example, a [Smithy Shape](https://smithy.io/2.0/spec/model.html#shapes), represented by a node in the ASG, would enjoy edges to each of its member shapes.
```
[the remark this answers is not in the document]
Comment (david-perez) [reply design/src/rfcs/rfc0034_generating_reusable_code.md:13]: Hence I'd remove the AST definition entirely, since it's not referred to in the body of the RFC; it's only used to introduce the ASG concept in a rather confusing way.

**Review by david-perez — COMMENTED** [review]
In _general_, I agree with the sentiment of the RFC: adherence to these 3 principles will not only make the generated Rust code more directly map to the Smithy model and be more reusable, but the code generator's code itself will become more composable, easier to understand, and as a result more maintainable.

However, I also think we have to accommodate for exceptions to these rules. For example:

1. `OperationNormalizer` is important for client SDKs to ensure backwards compatibility, which this text deems a violation of principle (2).
2. Doing https://github.com/awslabs/smithy-rs/issues/2363 would violate principle (1).

In other words, I don't care as much about "purity" if it makes our lives harder in honoring the full implications of the Smithy spec, or if we hamper UX.

File: design/src/rfcs/rfc0034_generating_reusable_code.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (hlbarber) [reply design/src/rfcs/rfc0034_generating_reusable_code.md:13]: I included it because I wanted to piggyback off the definition a little and it is mentioned briefly to say that 

> At high-level, the [Smithy library](https://github.com/awslabs/smithy-rs) library provides not only a way to parse the AST defined by the...

But yeah, I agree this has turned out badly. I will try again with concrete references to what it means for Smithy.

File: design/src/rfcs/rfc0034_generating_reusable_code.md [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+        /* Deserialize using from the "/cat/{type}" URI */
+    }
+}
+```
+
+If we removed the name aliasing, and named the Rust type `Food` in both cases, we'd then encounter a conflicting implementation.
+
+In contrast, the `FromRequest` implementations generated from `RestJsonPetService > Dog > Food` and `AwsJsonPetService > Dog > Food` are compatible:
+
+```rust
```
[the remark this answers is not in the document]
Comment (hlbarber) [reply design/src/rfcs/rfc0034_generating_reusable_code.md:122]: They are compatible in the sense that `impl FromRequest<RestJson1>` can co-exist with `impl FromRequest<AwsJson1_0>`. Whereas the snippet above is incompatible because they're both `impl FromRequest<RestJson1>` with non-identical implementations

[the remark this answers is not in the document]
Comment (hlbarber) [reply design/src/rfcs/rfc0034_generating_reusable_code.md:122]: There's probably a whole section about how Rusts coherence rules play nice with all the principles. For example, if you generate for example the `OperationShape` ZST in the `operationShape` method in alignment with (3) and the a "`ServiceShape`" ZST in the `serviceShape` method then you can always impl `FromRequest<Service, Op>` without conflicts - even when you had generated the service, operation and input in separate crates.

File: design/src/rfcs/rfc0034_generating_reusable_code.md [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+impl FromRequest<AwsJson1_0> for DogInput {
+    async fn from_request(request: http::Request) -> Self {
+        /* Deserialize ignoring the URI entirely */
+    }
+}
+```
+
+This is because `FromRequest` is parameterized by the protocol and therefore a `FromRequest<RestJson1>` can co-exist with a `FromRequest<AwsJson1_0>`.
+
+<!-- Interesting aside: If you have a recursive structure which includes an input then the code generator maps a single shape to two distinct types within one service closure. -->
```
[the remark this answers is not in the document]
Comment (hlbarber) [reply design/src/rfcs/rfc0034_generating_reusable_code.md:138]: This is a shower thought - I should remove it. The odd case is supposed to be 1 service, 1 operation, 1 input structure and **no** other shapes can still make the code generator map 1 input structure to 2 Rust types in the case of recursion.

File: design/src/rfcs/rfc0034_generating_reusable_code.md [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+Why are the generated types for `RestJsonPetService > Dog > Food` and `RestJsonPetService > Cat > Food` incompatible? How should be classify such failures?
+
+### Directed Closures
+
+A (directed) closure, `closure(S)`, of a given shape `S` is all shapes visitable by downwards directed traversal. In our example model `closure(Food)` only contains `Food`, whereas `closure(RestJsonPetService)` contains everything except for `AwsJsonPetService`.
```
[the remark this answers is not in the document]
Comment (hlbarber) [reply design/src/rfcs/rfc0034_generating_reusable_code.md:144]: Whoops. Nice catch.

File: design/src/rfcs/rfc0034_generating_reusable_code.md (hunk 1 above)
Comment (82marbag) [reply design/src/rfcs/rfc0034_generating_reusable_code.md:3]: Sorry. That's what I mean. Update the "status" field before merging to either "Accepted" or "Rejected" depending on the decision. My point is to start as `Status: RFC`

File: design/src/rfcs/rfc0034_generating_reusable_code.md [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+> Visiting any valid model should produce valid code.
+
+Notice that without complimentary principles this is not particularly strong requirement - a code generator which produces an empty crate for all models is sufficient. All proposed three principles should be used together to make decisions.
+
+## Principle 2) Only Traverse Downwards
+
+> The `ShapeVisitor` methods, if provided with access to the model, should not traverse upwards.
+
+The motivation for this principle is that adherence is sufficient to ensure the following formal description of type reusability:
```
[the remark this answers is not in the document]
Comment (hlbarber) [reply design/src/rfcs/rfc0034_generating_reusable_code.md:202]: What I'm trying to convey here is the following hand wavey proof:

We have that `S = (S \ (S ∩ T)) ∪ (S ∩ T)` and all `S \ (S ∩ T)` are _above_ `S ∩ T`. This means that when we generate `S` and the visitor gets to `S ∩ T`, because we can't traverse upwards and inspect `S \ (S ∩ T)` we MUST act equivalently to the case where we're generating `T` (for which the same is true). Therefore it's (2) => `S ∩ T` is compatible when generating `S` and `T`.

The point of (2) is that we're not supposed to have any state concerning the parents to operate on. Maybe I should state this more explicitly? For example, having the `ShapeVisitor` record the name of the service as it traverses down the ASG is technically not a violation of (2), but is supposed to be.

File: design/src/rfcs/rfc0034_generating_reusable_code.md (hunk 1 above)
Comment (LukeMathWalker) [reply design/src/rfcs/rfc0034_generating_reusable_code.md:3]: Either works to be honest, one can start optimistic and then change to `Rejected` before merging 😛 I have no strong opinion there.

**Comment by jdisanti** [comment]
A high level comment: this needs to consider how things will work in a Smithy `DirectedCodegen` future since that's the direction Smithy code generators are headed in. The `ShapeVisitor` approach is currently not the path forward.

File: design/src/rfcs/rfc0034_generating_reusable_code.md [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
+> The code generator should be such that, for any models `S` and `T`, the types in the intersection, `S ∩ T`, should be compatible.
+
+Why is it sufficient? Any types in the generated `S ∩ T` cannot depend on the details of their parents and so visiting `S ∩ T` in the while traversing `S` and `T` should be identical.
+
+Notice that it is a special case of the "model crate proposal". A less obvious consequence is that expanding a model under this rule causes a non-breaking change.
```
[the remark this answers is not in the document]
Comment (hlbarber) [reply design/src/rfcs/rfc0034_generating_reusable_code.md:208]: The proof goes something like this:

Suppose `N` is a submodel of `M` - this is what I mean by expansion. We have that `M = (M \ N) ∪ N`. `M \ N` is not below `N`, if it was then `N` would not be a valid model. So visiting `N` for both model `N` and `M` should result in identical code, which means that any visiting `M \ N` must result in additional code (i.e. a non-breaking change).

**Comment by hlbarber** [comment]
> this needs to consider how things will work in a Smithy `DirectedCodegen` future since that's the direction Smithy code generators are headed in. . The ShapeVisitor approach is currently not the path forward.

This is a good question. I opted to reference the `ShapeVisitor` to make an abstract proposal a bit more concrete, the language of traversal upwards/downwards _should_ still make sense in `DirectedCodegen` approach. I can attempt to translate it to the language of `DirectedCodegen` or make the proposal less concrete?

One concern to me is that Smithy will not provide a first class interface to traverse over anything but service closures. In which case we could never reap the large benefits of this proposal.

**Comment by hlbarber** [comment]
The ability to reuse types has is not a priority for this year and as a result I will not push this out of draft. I will continue to respond to comments.
